//
//  circuit_breaker.proto
//  Progetto di Distributed Systems and Big Data
//  Anno Accademico 2024-25
//  (C) 2024 Luca Montera, Alessio Giordano
//
//  Created by Alessio Giordano on 21/11/24.
//

syntax = "proto3";
import "http.proto";

service CircuitBreaker {
	rpc send(CircuitBreakerRequest) returns (CircuitBreakerResponse);
}

enum CircuitBreakerStatus {
    CircuitBreaker_CLOSED = 0;
    CircuitBreaker_HALF_OPEN = 1;
    CircuitBreaker_OPEN = 2;
}

message CircuitBreakerRequest {
    string id = 1;
    repeated HTTPStatusCode expected = 2;
    uint32 timeout = 3; // HTTP request timeout
    uint32 threshold = 4; // Consecutive failures triggering the opening of the circuit
    uint32 recovery = 5; // Seconds to wait in open circuit
    HTTPRequest http = 6;
}

message CircuitBreakerResponse {
    string id = 1;
    CircuitBreakerStatus status = 2;
    uint32 wait = 3; // Seconds remaining in open circuit
    HTTPResponse http = 4;
}